optaplanner-docs/src/main/asciidoc/ConstraintStreams/ConstraintStreams-chapter.adoc

```diff
@@ -482,6 +483,58 @@ The following example finds a computer which runs the most power-demanding proce
 ----
 
 
+[[constraintStreamsConditionalPropagation]]
+=== Conditional propagation
+
+Conditional propagation enables you to exclude constraint matches from the constraint stream based on the presence or
+absence of some other object.
+The following example penalizes computers which have at least one process running on them:
+
+[source,java,options="nowrap"]
+----
+    private Constraint runningComputer(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(CloudComputer.class)
+                .ifExists(CloudProcess.class, Joiners.equal(Function.identity(), CloudProcess::getComputer))
+                .penalize("runningComputer",
+                        HardSoftScore.ONE_SOFT,
+                        computer -> ...);
+    }
+----
+
+Note the use of the `ifExists()` building block.
+On `UniConstraintStream`, the `ifExistsOther()` building block is also available, useful in situations where your
+`from()` constraint match type is the same as your `ifExists()` type.
+
+Conversely, the `ifNotExists()` (and `ifNotExistsOther()` on `UniConstraintStream`) are available to achieve the
+opposite effect:
+
+[source,java,options="nowrap"]
+----
+    private Constraint unusedComputer(ConstraintFactory constraintFactory) {
+        return constraintFactory.from(CloudComputer.class)
+                .ifNotExists(CloudProcess.class, Joiners.equal(Function.identity(), CloudProcess::getComputer))
+                .penalize("unusedComputer",
+                        HardSoftScore.ONE_HARD,
+                        computer -> ...);
+    }
+----
+
+Here, only the computers without processes running on them will be penalized.
+
+Also note the use of the `Joiner` class to limit the constraint matches.
+For a description of `Joiners`, see <<constraintStreamsJoin,joining>>.
+Conditional propagation operates much like joining, with the exception of not increasing the
+<<constraintStreamsCardinality, stream cardinality>>.
+The matches from `ifExists()` and other similar building blocks will not be available further down the stream.
+
+[NOTE]
+====
+For performance reasons, using conditional propagation with the appropriate `Joiner` is preferable to joining.
+Joining should only be used in cases where the other instance is actually required for another operation further down
+the stream.
```

Comment:
I'd argue that's not the real difference between join() and ifExists().
This is the real difference:
[a1, a2, a3].join([b1, b2], someCondition) gives up to 6 matches.
[a1, a2, a3].ifExists([b1, b2], someCondition) gives up to 3 matches.

Reply:
Well, that's the same thing I'm trying to say. Obviously, I'm not saying it well enough.